Ensure queue isn't empty before trying to get a new module

baurmatt opened the pull request:
Fixes #940.

baurmatt committed:
Ensure queue isn't empty before trying to get a new module

Fixes #940.

reidmv committed:
Revert "Log error message on rescued ThreadError"

This reverts commit e00a2e3304335479c389204d8518664c0cabb54e.

Fixes #941
Fixes #940

Per the docs on Queue:

> If the queue is empty, the calling thread is suspended until data
> is pushed onto the queue. If non_block is true, the thread isn't
> suspended, and ThreadError is raised.

So, a ThreadError occurs very normally, when the queue is empty. This is
how threads are supposed to terminate. Rather than warn on this
occurance it sounds like we should just debug.

Sharpie commented on lib/r10k/puppetfile.rb:
```diff
@@ -195,7 +195,7 @@ def visitor_thread(visitor, mods_queue)
         while mod = mods_queue.pop(true) do mod.accept(visitor) end
       rescue ThreadError => e
-        logger.error _("Thread error during concurrent module deploy: %{message}") % {message: e.message}
+        logger.debug _("Module thread %{id} exiting: %{message}") % {message: e.message, id: Thread.current.object_id}
```
After re-reading the docs a bit, it looks like `ThreadError` is raised by `.pop(true)` when the queue is empty. Logging this at DEBUG level seems reasonable. However, we should add an additional rescue clause at the end of the block to catch and log errors that occur during `mod.accept`:

```ruby
rescue => e
  logger.error _("Error during concurrent deploy of module %{mod_name}: %{message}") %
    {mode_name: mod.name,
     message: e.message}
  Thread.exit
end
```

reidmv replied:
Added additional rescue block

reidmv committed:
Catch unexpected mod.accept errors

For better exception tracking and logging

Sharpie reviewed (approved):
Looks good 👍

reidmv committed:
Improve exception handling for threaded deploy

If a module thread raises a non-ThreadException, the behavior should
roughly match what serial deployment does. The exception should cause
r10k to halt the deploy.

This commit implements that behavior, while also ensuring that we don't
interrupt in-progress threads that have not raised an exception
themselves. Instead we wait for threads to finish their current module
before they halt and the main thread re-raises the first thread
exception encountered.
